강원대_Android_이아림_6주차과제_STEP2

arieum opened the pull request:
### 💁‍♀️Foreground_Custom_Notification 실행영상
https://github.com/user-attachments/assets/cbaa5e66-879f-4472-b525-3234f5f47267

### 💁‍♀️Background_Notification 실행영상
https://github.com/user-attachments/assets/e6b75d2e-3d0f-4788-b835-6f1364347a25

arieum committed:
Resolved conflicts & Add google-services.json(.gitignore)

arieum committed:
Feat: Write README.md

arieum committed:
Create: Add initial entry screen_WelcomeActivity

arieum committed:
feat: Add Remote_Config function

arieum committed:
refactor: databinding 제거

arieum committed:
refactor: Resolve Issue!

arieum committed:
refactor: Resolve issue(Add coroutine)

arieum committed:
refactor: Add WelcomeViewModel_init_function

arieum committed:
feat: Add Foreground_Custom_Notification

arieum committed:
feat: Add Foreground_Custom_Notification

arieum committed:
feat: Add Function_firebase_MSG_Recieved

arieum committed:
feat: Add Function_firebase_MSG_Recieved

arieum committed:
refactor: remove unnecessary import & typo correction

arieum committed:
refactor: remove unnecessary import & typo correction

arieum committed:
Delete MyPlaceContract.kt

arieum committed:
Delete PlaceDbHelper.kt

arieum committed:
Delete RetrofitClient.kt

arieum committed:
refactor: Resolved conflict in .gitignore

arieum committed:
refactor: Resolved conflicts in gradlew

arieum committed:
refactor: Resolved conflicts in gradlew.bat

arieum committed:
refactor: Inject sharedPreference in MapRepository.kt

arieum committed:
refactor: changed mapRepository parameter in RepositoryModule.kt

arieum committed:
refactor: add const variable KakaoAK in PlaceRepository.kt

arieum committed:
refactor: Apply STEP1 Feedback

arieum committed:
refactor: Apply STEP1 Feedback

arieum committed:
refactor: remove img entity in place

arieum committed:
refactor: remove img entity in place

mkSpace commented:
리베이스 해주시고 Repository Pattern CodeLab 링크 드리니 참고해서 변경해보세요

https://developer.android.com/codelabs/basic-android-kotlin-training-repository-pattern?hl=ko#0

arieum committed:
Merge branch 'arieum' into arieum_step2

arieum committed:
refactor: Replace runBlocking with coroutineScope to prevent ANR

arieum committed:
Merge remote-tracking branch 'arieum_remote/arieum_step1' into arieum_step1

arieum committed:
rebase

arieum committed:
Resolved conflicts

arieum commented:
저 뭘 잘못했길래... STEP1 멘토님 코멘트가 사라진거죠 .... 😢

![image](https://github.com/user-attachments/assets/c74d602d-7d0f-4ff3-9f51-60afc17d1de4)

rebase가 잘못된걸까요?? 제가 직접 닫지 않았습니다 !!

arieum committed:
refacotr: typo correction

arieum commented:
~~⚠️ 수정 중입니다 ⚠️~~

💁‍♀️In Logviewmodel, **runBlocking**을 없애고 **viewmodelscope**으로 바꾸었습니다
-  runBlocking은 자칫 잘못하면 ui스레드를 오랫동안 블로킹하여 anr를 발생시킬 수 있는 여지가 있다는 점을 알고, viewmodelscope 코루틴으로 바꿨습니다

💁‍♀️**suspendCoroutine** 관련 ···
-  remoteConfig 값을 가져오고 난 후에 true/false 값을 반환할 수 있는 도구를 찾는 도중 `suspendCoroutine`이라는 걸 봤습니다...! 그치만 불필요한 걸까요...?
- https://velog.io/@minlove2013/Kotlin-%EB%B9%84%EB%8F%99%EA%B8%B0-%EB%B0%A9%EC%8B%9D%EC%9C%BC%EB%A1%9C-%EB%8D%B0%EC%9D%B4%ED%84%B0%EB%A5%BC-%EC%B2%98%EB%A6%AC%ED%95%A0-%EB%95%8C-%EC%9B%90%ED%95%98%EB%8A%94-%EA%B0%92%EC%9D%B4-%EB%82%98%EC%98%A4%EC%A7%88-%EC%95%8A%EC%95%84-SuspendCoroutine-%EC%9D%B4%EC%9A%A9
- 마지막 수정까지 저는 이걸 냅두기로 결정했습니다..! `await`로 결과값을 반환할 수도 있지만 이 코루틴이 더 명시적으로 보입니다! 

💁‍♀️`rebase`에 시간을 쓸데없이 너무 많이 써서.. (울고시펐슴민다...🥲그렇게 시간을 많이 쓴 결과가 이 모양입니다ㅠ죄송합니다ㅠ이번에 처음으로 rebase 쓴것같습니다....)  공유해주신 링크 이제 보고 있습니다...! 읽고 낼 아침전까지 다 수정하겠습니다 ★

💁‍♀️In welcomeViewActivity, map화면으로 이동 하기전 delay(1s)하기 위해 **lifecyclescope.launch{}**를 적용했습니다
- delay( ) 중단 함수를 추가하기 위해 일단 lifecyclescope를 선언하긴 했는데... 
- `"1초지연하고 -> 지도화면으로 이동 "` 이 액션을 위해서 액티비티가 파괴되야 취소되는 lifecyclescope 사용은 너무 과한걸까요? 
- 흠.. 차라리 아래 코드가 나았을까요??

```
Handler(Looper.getMainLooper()).postDelayed({moveMapView()}, 1000)
```

💁‍♀️지금 제 코드는 viewmodel에 비지니스 로직이 많이 섞여있고, repository와 manager 클래스 간에 혼용이 있어서 재사용성이 떨어져보입니다.. 멘토님이 공유해주신 링크를 참고하라한 것도 이때문이었을 것 같습니다. 어제 나름 열심히 수정했다고 생각했는데 오늘 다시 보니 논점을 잘못 짚은것 같아요..

arieum committed:
refactor: Use Dispatcher.IO for database operations in LogRepository

arieum committed:
refacotr: Restructure package

arieum committed:
refactor: refactor db&repository implementation

- simplify Appdatabase class by removing companion object
- update databaseModule to directly build database instance
- modify logrepo to use appdatabase instead of placedao
- remove explicit dispatcher.io usage

arieum committed:
refactor: Improve delayed UI update using coroutines in WelcomeAcitivity

mkSpace commented on app/src/main/java/campus/tech/kakao/map/presentation/view/WelcomeActivity.kt:
```diff
@@ -0,0 +1,103 @@
+                    // 사용자에게 권한 요청 이유를 설명하는 UI
+                    AlertDialog.Builder(this)
+                        .setTitle("알림권한이 필요합니다")
+                        .setMessage("This app needs notification permission to send you updates and important information")
```
이런 문자열은 strings.xml로 다 빼주세요

mkSpace reviewed (commented):
아림님 6주 동안 고생 많으셨습니다.

최종적으로 작성하신 코드들 보니 확실히 많이 깔끔해졌다는게 보이네요.

다만 Hilt, Room, Coroutine을 사용한 프로젝트는 거의 처음이시다 보니 아직 어색한 부분은 다소 보이는 것 같습니다.

이 부분은 너무 당연한 부분이니 제가 이전에 링크 드렸던 코드랩이나 여러 다른 잘 만든 레포들 보고 학습하시면 금방 실무와 가까운 코드를 작성하실거라 믿습니다.

리뷰 남겨드리니 확인 부탁드려요!

---

지금 현재 AndroidManifest에 작성하신 대로 Notification을 수신 받긴 할 텐데 기능 요구사항 중에 foreground 상태에서 수신 시 커스터마이징을 하라는 요구사항이 있어서요. 이게 제대로 동작하는지는 잘 모르겠네요.

이 부분을 커스터마이징 하려면 보통 FirebaseMessagingService에서 Background로 수신 받고 Notification을 띄우는 형태가 일반적인데 이를 활용한 부분은 누락되어 있는것 같아서 여쭤봅니다

mkSpace commented:
> ~⚠️ 수정 중입니다 ⚠️~
> 
> 💁‍♀️In Logviewmodel, **runBlocking**을 없애고 **viewmodelscope**으로 바꾸었습니다
> 
> * runBlocking은 자칫 잘못하면 ui스레드를 오랫동안 블로킹하여 anr를 발생시킬 수 있는 여지가 있다는 점을 알고, viewmodelscope 코루틴으로 바꿨습니다
> 
> 💁‍♀️**suspendCoroutine** 관련 ···
> 
> * remoteConfig 값을 가져오고 난 후에 true/false 값을 반환할 수 있는 도구를 찾는 도중 `suspendCoroutine`이라는 걸 봤습니다...! 그치만 불필요한 걸까요...?
> * https://velog.io/@minlove2013/Kotlin-%EB%B9%84%EB%8F%99%EA%B8%B0-%EB%B0%A9%EC%8B%9D%EC%9C%BC%EB%A1%9C-%EB%8D%B0%EC%9D%B4%ED%84%B0%EB%A5%BC-%EC%B2%98%EB%A6%AC%ED%95%A0-%EB%95%8C-%EC%9B%90%ED%95%98%EB%8A%94-%EA%B0%92%EC%9D%B4-%EB%82%98%EC%98%A4%EC%A7%88-%EC%95%8A%EC%95%84-SuspendCoroutine-%EC%9D%B4%EC%9A%A9
> * 마지막 수정까지 저는 이걸 냅두기로 결정했습니다..! `await`로 결과값을 반환할 수도 있지만 이 코루틴이 더 명시적으로 보입니다!
> 
> 💁‍♀️`rebase`에 시간을 쓸데없이 너무 많이 써서.. (울고시펐슴민다...🥲그렇게 시간을 많이 쓴 결과가 이 모양입니다ㅠ죄송합니다ㅠ이번에 처음으로 rebase 쓴것같습니다....) 공유해주신 링크 이제 보고 있습니다...! 읽고 낼 아침전까지 다 수정하겠습니다 ★
> 
> 💁‍♀️In welcomeViewActivity, map화면으로 이동 하기전 delay(1s)하기 위해 **lifecyclescope.launch{}**를 적용했습니다
> 
> * delay( ) 중단 함수를 추가하기 위해 일단 lifecyclescope를 선언하긴 했는데...
> * `"1초지연하고 -> 지도화면으로 이동 "` 이 액션을 위해서 액티비티가 파괴되야 취소되는 lifecyclescope 사용은 너무 과한걸까요?
> * 흠.. 차라리 아래 코드가 나았을까요??
> 
> ```
> Handler(Looper.getMainLooper()).postDelayed({moveMapView()}, 1000)
> ```
> 
> 💁‍♀️지금 제 코드는 viewmodel에 비지니스 로직이 많이 섞여있고, repository와 manager 클래스 간에 혼용이 있어서 재사용성이 떨어져보입니다.. 멘토님이 공유해주신 링크를 참고하라한 것도 이때문이었을 것 같습니다. 어제 나름 열심히 수정했다고 생각했는데 오늘 다시 보니 논점을 잘못 짚은것 같아요..

제가 질문 형태로 리뷰 드렸던 부분은 상당 부분 알고 사용하시는지 궁금해서 여쭤본 거였구요. 해당하는 코드가 정말 자신이 생각하는대로 동작하는지 제대로 따져봤음 해서 리뷰 드린거였어요.

코드를 작성하실 때 여러 블로그 등을 참조하실 텐데 그 코드들이 과연 제대로 동작하는가? 현재 상황에서 최선의 방법인가? 하는 의문을 항상 갖고 작성하신다면 좋은 코드를 작성하시는데 도움이 될거라 생각이 듭니다.

arieum committed:
refactor: refactor FCMservice for proper initialization and foreground notification handling

arieum committed:
refactor: Move hardcoded strings to string resources

arieum commented:
### 최종 수정본 이후 실행영상입니다!
Foreground일때도 firebasemessage를 수신할 수 있도록 수정했습니다! 
(제가 과제를 잘못 이해했던것같아요.. foreground일때는 firebasemessage없이도 notification이 울리도록 하는 건 줄 알았습니다..🥲) 

https://github.com/user-attachments/assets/52e0302a-cf51-4ed3-9669-4e717bbbb276

arieum commented:
안녕하세요! 끝인사가 늦었습니다 하핳🥲 
마지막 피드백에 대해 코드 수정했습니다!! 확인해주시면 감사하겠습니다!! 
정확하고 깔끔한 코드로 수정하고 싶어서 지금도 열심히 구글링하고 있습니다!!

한 달 동안 정말 감사했습니다! 제가 멘토님을 정말 귀찮게 했는데도 항상 친절하게 도와주셔서 감사했습니다🙇‍♀️ 
멘토님과 컴퓨터 화면을 두고 나눈 대화들이 저한텐 이 난폭한 커리큘럼을 버틸 수 있는 든든한 지지대 같았습니다!😊 
제 딴에는 멘토님의 피드백으로 나름 많이 성장했다고 생각하는데.. 멘토님 입장에서는 제가 이제야 겨우 들숨 날숨을 해내는 걸로 보이시겠죠😭?

이 프로젝트가 제 첫 프로젝트라서 애장품같아요 q(≧▽≦q)!! 앞으로도 계속해서 진화시킬 생각입니다! 나중에 번듯한 앱이 되면 가장 먼저 멘토님께 보여드리고 싶습니다😆!! 

멘토님은 더이상 이 하찮은 멘티를 보고 싶지 않으시겠지만..ㅠㅠ 저는 너무너무 아쉽습니다🥲
나중에 기회가 된다면 꼭 뵙고싶습니다!! 🙇‍♀️🙇‍♀️ 
정말 감사했습니다 멘토님!!☺️
![JigglypuffPokemonGIF](https://github.com/user-attachments/assets/781c4370-0d7e-4961-acf4-c00f4928e995)
